### ISSUE TITLE
Strange error when dealing with `Something[NoReturn]` as a type parameter

### ISSUE THREAD

**Issue by arseniiv** [python/mypy#8266]
The error in this code seems to be a bug:

```python
from typing import List, NoReturn, Sequence

a: List[List[int]] = [] # ok
b: List[NoReturn] = [] # ok
c: List[List[NoReturn]] = [[]] # ok
d: List[List[NoReturn]] = [[]][:0] # ok
e: List[List[NoReturn]] = [] # error
f: Sequence[List[NoReturn]] = [] # ok
```

Error message:

> Incompatible types in assignment (expression has type "`List[<nothing>]`", variable has type "`List[List[NoReturn]]`")

In addition, one can replace `NoReturn` here with any other type, and the error disappears.

Python 3.8, mypy 0.761 (from PyPI), no flags. (I use it all through VS Code but that shouldn’t be an issue.)

**Comment by JelleZijlstra** [comment]
`NoReturn` is treated as the bottom type, which means there are no instances of it, and it is meaningless to have a `List[NoReturn]`. What is your use case for this?

**Comment by arseniiv** [comment]
@JelleZijlstra Why is it meaningless? I can construct an empty list of this type, and if for some reason I would want to specify that a function returns only `[]`, I could use this type if I’ll find myself lazy to use `List[T]` for some typevar `T`.

In my real case, though, I use `NoReturn` as a phantom type. Let’s say we have this:
```python
T = TypeVar('T', covariant=True, bound='B')

class A(ABC, Generic[T]): ...

class B(A[B]): ... # no A, B or C mentioned inside

class C(A[T]):
    children: Sequence[A[T]]
    ...
```
This says that `A[B]` may be a `B` instance or something with `B` instances inside (or without); but `A[NoReturn]` (which is neatly aliased like `AWithoutB`) can be only an instance of `C` which moreover can contain in `children` just instances of `C` with the same property—that is, no `B` anywhere inside. (By the way, the interface of `A` makes use of `T` in types of several methods, so there’s no merit to simplify the overall construction.)

This code, or at least the original one, typechecks nicely and works as expected too—up until I do something like
```python
things: List[Tuple[str, AWithoutB]] = []
```
(`List` is to allow appending something later) and the error happens. At least it’s fixable with `cast`. :slightly_smiling_face: 

Anyway, the former code illustrates there’s an inconsistency: `d` would be an empty list just like `e` but one is allowed and the other isn’t; I think this is clearly something to be fixed even if those types have no good use cases; just maybe with low low priority.

**Comment by Zabolekar** [comment]
> it is meaningless to have a `List[NoReturn]`

0. This hardly explains the current behavior: the example contains multiple `List[NoReturn]`, but only one of them causes problems.

1. What should be the type of `x` in the following snippet? (I do admit that it doesn't look useful)

```python
from typing import NoReturn

def f() -> NoReturn:
    raise Exception

x = [f()]
```

(the revealed type is `builtins.list[[<nothing>]`, but I couldn't find whether there is a difference between `NoReturn` and `<nothing>`)

**Comment by msullivan** [comment]
That's very clever. And it works and will actually catch the errors you are trying to?

Yeah, I think they are the same.

I'd accept a patch that fixes this, but it seems low-priority otherwise.

**Comment by arseniiv** [comment]
@msullivan Yeah, I hope it does! Tested it on several expressions, and the messages in case of mismatch were helpful.

> I'd accept a patch that fixes this, but it seems low-priority otherwise.

Thanks! I’ll try to make myself familiar with mypy code after a while, then. Maybe I’ll understand what should be changed, then we could hopefully have a patch.

**Comment by arseniiv** [comment]
If someone’s interested in the status, I’ve read in a lazy manner a couple of source files I thought may be related, but haven’t found the likely cause yet.

**Comment by ilevkivskyi** [comment]
The original example now passes without errors on current master.
